---
title: "Class 11: Structural Bioinformatics pt.2"
author: "Serena Quezada"
format: pdf
---

## AlphaFold DB

The EBI maintains the largest database of Alpha fold structure prediction models at: https://alphafold.ebi.ac.uk

From last class we saw that the PDB had 244,290 (Oct 2025)

The total number of protein sequences in UniProtKB is 199,579,901

>Key Point: this is a tiny fraction of sequence space that has structural coverage (0.12%)

```{r}
244290/199579901 * 100
```
AFBD is attempting to address this gap...

There are two "Quality Scores" from AlphaFold one for residues (i.e. each amino acid) called **pLDDT** score. The other **PAE** score measures the confidence in the relative position of two residues (i.e. a score for every pair of residues). 

## Generating your own structure predictions 

picture 

![](hivprdimer_23119_001.png)
![](hivprdimer_23119_002.png)
chain A is being superimposed


![](hivprdimer_23119_004.png)
chain B is being superimposed 


![](hivprdimer_23119_005.png)

# Custom analysis of resulting models in R

Read key result files into R. The first thing I need to know is what my results directory/folder is called (i.e. it name is different for every AlphaFold run/job)

```{r}
results_dir <- "HIVPR_dimer_23119"
```

```{r}
pdb_files <- list.files(path=results_dir,
                        pattern="*.pdb",
                        full.names = TRUE)

basename(pdb_files)
```
```{r}
library(bio3d)

pdbs <- pdbaln(pdb_files, fit=TRUE, exefile="msa")

```
```{r}
pdbs
```

```{r}
rd <- rmsd(pdbs, fit=T)

range(rd)
```
```{r}
pdb <- read.pdb("1hsg")
```

```{r}
plotb3(pdbs$b[1,], typ="l", lwd=2, sse=pdb)
points(pdbs$b[2,], typ="l", col="red")
points(pdbs$b[3,], typ="l", col="blue")
points(pdbs$b[4,], typ="l", col="darkgreen")
points(pdbs$b[5,], typ="l", col="orange")
abline(v=100, col="gray")
```
```{r}
core <- core.find(pdbs)
```
```{r}
core.inds <- print(core, vol = 0.5)
```

```{r}
xyz <- pdbfit(pdbs, core.inds, outpath="corefit_structures")
```

```{r}
rf <- rmsf(xyz)

plotb3(rf, sse=pdb)
abline(v=100, col="gray", ylab="RMSF")
```

# Residue conservation from alignment file

```{r}
aln_file <- list.files(path=results_dir,
                       pattern=".a3m$",
                        full.names = TRUE)
aln_file
```

```{r}
aln <- read.fasta(aln_file[1], to.upper = TRUE)
```

```{r}
 dim(aln$ali)
```

```{r}
sim <- conserv(aln)

plotb3(sim[1:99], sse=trim.pdb(pdb, chain="A"),
       ylab="Conservation Score")
```

