---
title: "Class 12: RNA Seq analysis"
author: "Serena Quezada (PID:A18556865)"
format: html
toc: true
---

## Background 
Today we will analyze some RNASeq data from Hime et al. on the effects of a common steroid (dexamethasone) on airway smooth muscle cells (ASM cells).

Our starting point is the "counts" data and metadata the count value for each gene in their different experiments (i.e. cell lines with or without the drug).

## Data Import 

```{r}
# Complete the missing code 
counts <- read.csv("airway_scaledcounts.csv", row.names = 1)
metadata <- read.csv("airway_metadata.csv")
```

```{r}
head(counts)
```

> Q1. How many genes are in this dataset?

```{r}
nrow(counts)
```

> Q. How many different experiments (columns in counts or rows in metadata are there?)

This is the amount of experiments 
```{r}
nrow(metadata)
```

```{r}
head(metadata)
```

> Q2. How many ‘control’ cell lines do we have? 

```{r}
sum(metadata$dex == "control")
```

## Toy differential gene expression

To start analysis let's calculate the mean counts for all genes in the "control" experiments.

1. Extract all "control" columns from the `counts` object 
2. Calculate the mean for all rows (genes) of these "control" columns 

3-4. Do the same for the "treated"
5. Compare the `control.mean` and `treated.mean` values.

> Q3. How would you make the above code in either approach more robust? Is there a function that could help here? 

Below is how I would make the code more robust, considering this a better function to help find the mean values of both control and treated values. 

This extracts all the "controls" 
```{r}
control.inds <- metadata$dex == "control"
control.counts <- counts[ , control.inds]
```

This is calculating the means for these `control` means 
```{r}
control.means <- rowMeans(control.counts)
```

> Q4. Follow the same procedure for the treated samples (i.e. calculate the mean per gene across drug treated samples and assign to a labeled vector called treated.mean)

Now we are calculating the treated columns 
```{r}
treated.inds <- metadata$dex == "treated"
treated.counts <- counts[ , treated.inds]
```

```{r}
treated.means <- rowMeans(treated.counts)
```

Now we are comparing the control and treated mean values (mean value per gene)
```{r}
meancounts <- data.frame(control.means, treated.means)
head(meancounts)
```

Now make a plot for our values 

> Q5 (a). Create a scatter plot showing the mean of the treated samples against the mean of the control samples.

> Q5 (b). You could also use the ggplot2 package to make this figure producing the plot below. What geom_?() function would you use for this plot?

You would use the geom_point() to produce a plot 

```{r}
library(ggplot2)
ggplot(meancounts, 
       aes(x = control.means, y = treated.means)) +
       geom_point() + theme_bw()
```

> Q6. Try plotting both axes on a log scale. What is the argument to plot() that allows you to do this?

Make this a log plot, by doing this we have more plots and a direct dark linear line appears

```{r}
plot(meancounts, log="xy")
```

We often talk metrics like "log2 fold-change"

```{r}
# control/treated
log2(20/10)
```

```{r}
log2(40/10)
log2(10/40)
```

Let's calculate the log2 fold for our treated over control mean counts 
We use the log2 fold to help us identify and interpret differently expressed genes between control and treated genes
```{r}
meancounts$log2fc <- 
log2(meancounts$treated.means / 
       meancounts$control.means)
```

```{r}
head(meancounts)
```

A common "rule of thumb" is a log2 fold change cutoff of +2 and -2 to call genes "Up regulated" or "Down regulated". 

> Q8. Can you determine how many up regulated genes we have at the greater than 2 fc level?

```{r}
sum(meancounts$log2fc > +2, na.rm = T)

#na.rm=T means that it won't count the NA's
```

> Q9. Can you determine how many down regulated genes we have at the greater than 2 fc level?

```{r}
sum(meancounts$log2fc < -2, na.rm = T)
```

> Q10. Do you trust these results? Why or why not?

When using the log2fold we are missing statistical significance. We don't know if the control and treated are caused by random error * write more on what we're missing *

##DESeq2 Analysis 

Let's do this analysis properly and keep our inner stats nerd happy - i.e. are the differences we see between drug and no drug significant given the replicate experiments 

```{r, message=FALSE}
library(DESeq2)
```

For DESeq analysis we need 3 things

- count values (`contData`)
- metadata telling us about the columns in `countData` (`colData`)
- design of the experiment (i.e. what do you want to compare)

Our analysis function form DESeq2 will setup the input required for analysis by storing all these 3 things together

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = metadata,
                              design = ~dex)
```

The main function in DESeq2 that runs the analysis is called `DESeq()`

```{r}
dds <- DESeq(dds)
```

```{r}
res <- results(dds)
res
```

## Volcano plot

This is a common summary result figure from these types of experiments and plot the log2 fold-change vs the adjusted p-value. 

```{r}
plot(res$log2FoldChange, -log(res$padj))
abline(v = c(-2,2), col = "red")   #abline provides vertical cut-off lines 
abline(h = -log(0.04), col = "red") # provides cut-off horizontal lines 
```

## Save our results

```{r}
write.csv(res, file = "my_results.csv")
```

## Add gene annotation 

To help make sense of our results and communicate them to toher folks we need to add more annotation to our main `res` object. 

We will use two bioconductor packages to first map IDs to different formats including the classic gene "symbol" gene name. 

I will install these with the following commands:
`BiocManager::install("AnnotationDbi")`
`BiocManager::install("org.Hs.eg.db")`

```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)
```

Let's see what is in 'org.Hs.eg.db' with the `columns()` function:

```{r}
columns(org.Hs.eg.db)
```
We can translate or "map" IDs between any of these 26 databases using the `mapIDs()` function. 

```{r}

res$symbol <- mapIds(keys = row.names(res), # our current IDs
                     keytype = "ENSEMBL",   # the format of our IDs
                     x = org.Hs.eg.db,     # where to get the mappings from
                     column = "SYMBOL")    # the format/DB to map to 

head(res)
```

Add the mappings for "GENENAME" and "ENTREZID" and store as `res$genename` and `res$entrez`

```{r}
res$entrezid <- mapIds(keys = row.names(res), # our current IDs
                     keytype = "ENSEMBL",  # the format of our IDs
                     x = org.Hs.eg.db,     # where to get the mappings from
                     column = "SYMBOL")    # the format/DB to map to 
```


```{r}
res$entrezid <- mapIds(keys = row.names(res), # our current IDs
                     keytype = "ENSEMBL",   # the format of our IDs
                     x = org.Hs.eg.db,     # where to get the mappings from
                     column = "SYMBOL")    # the format/DB to map to 
res$genename <- mapIds(keys = row.names(res), # our current IDs
                     keytype = "ENSEMBL",   # the format of our IDs
                     x = org.Hs.eg.db,     # where to get the mappings from
                     column = "SYMBOL")    # the format/DB to map to 
```

```{r}
head(res)
```

## Pathway analysis 

There are lots of bioconductor packages to do this type of analysis. For now let's just try one called **gage** again we need to install this is we don't have it already.

```{r, message=FALSE}
library(gage)
library(gageData)
library(pathview)
```

To use **gage** I need two things 

- a named vector of fold-change values of our DEGs (our geneset of interest)
- a set of pathways or genesets to use for annotation. 

```{r}
x <- c("barry" = 5, "lisa" = 10)
x
```

```{r}
names(x) <- c("low", "high")
x
```

```{r}
foldchanges <- res$log2FoldChange
names(foldchanges) <- res$entrez
head(foldchanges)
```

```{r}
data(kegg.sets.hs)

keggres = gage(foldchanges, gsets = kegg.sets.hs)
```

In our results we have:
```{r}
attributes(keggres)
```

```{r}
head(keggres$less)
```
Let's look at one of these pathwats with our genes colored up so we can see the overlap

```{r}
pathview(pathway.id = "hsa05310", gene.data = foldchanges)
```

Add this pathway figure to our lab report 

![](hsa05310.pathview.png)
## Save our main results 

```{r}
write.csv(res, file = "myresults_annotated.csv")
```





